File: CDexhand/kEstimator.cpp
#include "kEstimator.hpp"

kEstimator::kEstimator()
{
    this->init();
}
void kEstimator::init()
{
    this->data_cnt = 0;
    this->k = 0.07;    
}

void kEstimator::data_push_in(double now_pos,double now_force)
{
    if(this->data_cnt == 0)
    {
        this->pos0 = now_pos-20;
        this->data_cnt++;
        now_pos = now_pos - this->pos0;
        this->theta << this->k,now_force-this->k*now_pos;

        this->phi << now_pos,1;
        this->P = this->phi * this->phi.transpose();
        this->P = this->P.inverse();
        this->last_pos = now_pos;
        return;
    }

    now_pos = now_pos - this->pos0;

    this->phi << now_pos,1;
    double err = now_force - (this->theta.transpose()*this->phi);
    double x = 130*(abs(err)-0.08);
    double lambda1 = abs(now_pos-this->last_pos);
    double lambda2 = x/(1.0+abs(x))+1.0;
    double forgot_factor = 1-lambda1*lambda2*1.575;
    if(forgot_factor < 0.3)
        forgot_factor = 0.3;

    this->K = (this->P*this->phi)/(forgot_factor + this->phi.transpose() * this->P * this->phi);
    this->P = ((Eigen::MatrixXd::Identity(2, 2) - this->K * this->phi.transpose()) * this->P ) / forgot_factor;
    this->theta = this->theta + this->K * err;
    this->k = this->theta(0);
    if(this->k < 0.01)
        this->k = 0.01;

    this->last_pos = now_pos;

}

File: pyDexhand/kEstimator.py
import numpy as np

class DexHand_kEstimator():
    PRELOAD_CNT = 20
    def __init__(self):
        self._data_cnt = 0 
        self._P = np.zeros((2,2))
        self._Phi = np.ones((DexHand_kEstimator.PRELOAD_CNT,2))
        self._y = np.ones((DexHand_kEstimator.PRELOAD_CNT))
        self._K = np.zeros((2,2))
        self.k = 0.07
        self._pos0 = -1
        self._minff = 1

    def data_push_in(self,nowpos,nowforce):
        if self._data_cnt == 0:
            self._pos0 = nowpos-20
        nowpos = nowpos-self._pos0

        if self._data_cnt < DexHand_kEstimator.PRELOAD_CNT:
            self._Phi[self._data_cnt][0] = nowpos
            self._y[self._data_cnt] = nowforce
            self._data_cnt += 1
            if self._data_cnt == DexHand_kEstimator.PRELOAD_CNT:
                self.theta = np.polyfit(self._Phi[:,0],self._y,deg=1)
                self._P = np.linalg.inv(self._Phi.T @ self._Phi)
                self._last_pos = nowpos
            return
        
        phi = np.array((nowpos,1))
        err = nowforce - np.dot(self.theta,phi)
        x = 130*(abs(err)-0.08)
        lambda1 = abs(nowpos-self._last_pos)
        lambda2 = x/(1+abs(x))+1
        forgot_factor = 1-lambda1*lambda2*1.575
        if forgot_factor < 0.3:
            forgot_factor = 0.3
        if forgot_factor < self._minff:
            self._minff = forgot_factor

        self._K = np.dot(self._P,phi)/(forgot_factor+np.dot(phi,np.dot(self._P,phi)))
        self._P = ((np.eye(2)-np.outer(self._K,phi)) @ self._P ) / forgot_factor
        self.theta = self.theta +self._K * err
        self.k = self.theta[0]
        if self.k<0.05:
            self.k = 0.05

        self._last_pos = nowpos